### ISSUE TITLE
[backend-compiler] fused sdsc program with >32 expert GEMM-chains fails the DDC scheduler (threshold 32-64)

### ISSUE THREAD

**Issue by ani300** [torch-spyre/torch-spyre#3769]
## Summary

The all-device MoE expert FFN is a per-expert loop of `x @ Wg[e]` → gelu-tanh SwiGLU → `@ Wd[e]`, accumulated into a device-resident `[T,H]` tensor. Compiling the **full 128-expert** body as a single fused `sdsc` program aborts in the DDC scheduler:

```
DtException: Incorrect chunk size in bytes.   L3DlOpsScheduler.cpp
sbf-run-scheduler-on-sdsc: failed on program 'sdsc_12'
dbo: pipeline failed
```

## Repro

Standalone, self-classifying, no model load / no real weights. Sweeps the number of experts fused into one compiled program across {8, 16, 32, 64, 128} at the real gemma-4-26B-A4B expert dims (D=2816, F=704) and reports the pass→abort boundary plus each compile's wall-time (to show the fast-fail). Lives in the hf-adapters MoE repro tree as `repros/gemma4_moe/issues/p6_expert_chain_scheduler_threshold.py`:

```python
import torch

D, F, T = 2816, 704, 8   # real gemma-4-26B-A4B expert dims (hidden, moe_intermediate)

def dense_expert_loop(x, gate, up, down):     # gate/up: [E,D,F]  down: [E,F,D]
    # == hf_gemma4_moe._moe_expert_chunk with router weight == 1; the Python
    # `for` unrolls into ONE fused sdsc program.
    acc = torch.zeros(T, D, dtype=x.dtype, device=x.device)
    for j in range(gate.shape[0]):
        a = torch.nn.functional.gelu(x @ gate[j], approximate="tanh") * (x @ up[j])
        acc = acc + (a @ down[j])
    return acc

for E in (8, 16, 32, 64, 128):
    x    = torch.randn(T, D, dtype=torch.float16)
    gate = torch.randn(E, D, F, dtype=torch.float16) * 0.02
    up   = torch.randn(E, D, F, dtype=torch.float16) * 0.02
    down = torch.randn(E, F, D, dtype=torch.float16) * 0.02
    torch.compile(dense_expert_loop, dynamic=False)(       # E >= 64 aborts
        x.to("spyre"), gate.to("spyre"), up.to("spyre"), down.to("spyre"))
```

The full script times each compile, verifies E≤32 against a CPU reference, and exits 0 iff a pass→abort boundary is observed in the sweep (i.e. the bug is still present).

## Hard threshold between 32 and 64 experts, not a timeout

Expert-count sweep, identical D-form body, shapes D=2816 / F=704:

| E (experts in one fused program) | Result |
|---|---|
| 8   | OK (34 s) |
| 16  | OK (67 s) |
| 32  | OK (142 s) |
| 64  | **ABORT (31 s)** |
| 128 | **ABORT** |

E=64 aborts **faster** (31 s) than E=32 **succeeds** (142 s) — so this is a hard scheduling threshold, not "takes longer and times out". The same shapes lower fine at E≤32, so it is not a shape/tensor bug; it is the number of expert GEMM-chains fused into one program.

## Why it matters to MoE

To keep the whole MoE FFN on device we must split the expert loop into ≤32-expert compiled regions. That split is what forces a **host-driven chunk loop**: ⌈128/32⌉ = 4 compiled-chunk dispatches per layer × 30 layers = **120 launch/sync round-trips per forward**. (The compute and the `[T,H]` accumulator stay on device; only the loop counter and buffer handoffs are on host.) A larger single-fused-program budget, or an on-device expert loop, would remove the host chunking and its dispatch overhead.

## Workaround currently in the adapter

Split the expert loop into `Ec ≤ 32`-expert compiled regions feeding a shared device `[T,H]` accumulator (`_MOE_EC=32`).

Epic: #3565

**Comment by ani300** [comment]
## No longer reproduces on current torch-spyre — closing

Re-ran the repro (`hf-adapters/repros/gemma4_moe/issues/p6_expert_chain_scheduler_threshold.py`, the dense per-expert loop == `_moe_expert_chunk` with router weight 1; D=2816, F=704, T=8) across the full E sweep:

| E   | result | compile time |
|-----|--------|--------------|
| 8   | OK (matches CPU) | 28s |
| 16  | compiles clean | 44s |
| 32  | compiles clean | 97s |
| 64  | compiles clean | 216s |
| 128 | compiles clean | 420s |

The `DtException: Incorrect chunk size ... L3DlOpsScheduler.cpp` **abort no longer occurs** — every E from 8 to 128 compiles cleanly, and compile time grows ~linearly with E (the opposite of the fast-fail scheduling threshold this issue described). The 32→64 DDC-scheduler threshold is gone. **Closing as fixed.**

### Caveat (tracked separately, not this issue)

Above E=8 the fused output **diverges from the float32 CPU reference, growing monotonically in E** (max|Δ| ≈ 0.06 / 0.09 / 0.16 / 0.30 at E=16/32/64/128, atol/rtol=2e-2 fp16 tol). A three-way precision A/B (device vs f32 ref vs fp16-accumulated ref vs the fp16-vs-f32 floor) shows the device runs ~2.5× above the pure fp16-accumulation floor and the max sits on near-zero output lanes — consistent with fp16 rounding compounded across the E-expert chain rather than a structural miscompile, but **not cleanly attributable to precision alone** at this tolerance. This is a separate accuracy concern from the scheduler abort and does not block closing #3769; flagging it here so it isn't lost.

Device layouts verified correct/consistent with the shipping `_MOE_EC=32` chunked path (accumulator feature-innermost `[H/64, T, 64]`; weights `[D, F/64, E, 64]` / `[F, H/64, E, 64]`).
